source/lambda_functionbsm_thea.py

```diff
@@ -0,0 +1,187 @@
+When a new file is added to the Sandbox s3 bucket, this lambda function will read
+the new JSON newline file, perform data transformation, and upsert the new data
+records to the corresponding Socrata data set on data.transportation.gov.
+
+Data transformation includes flattening of the data structure, which is
+required for the data to work with the Socrata backend on data.transportation.gov.
+Renaming of certain fields is done to achieve consistency across data sets. No
+unit conversion will be done to the data, though occasionally, additional fields
+will be added to enhance usage of the data set in Socrata (e.g. geolocation fields
+for plotting). Any added fields will be listed in the description of each Socrata
+data set.
+
+A separate lambda function using the lambda handler in `lambda_function_cleanup.py`
+will be used to keep the data set at a manageable size by removing oldest records
+from the Socrata data set.
+
+Requires:
+- Uploading to AWS requires packaging the source code with the Python extensions
+into a single zip file. This code requires the following Python libraries to be
+packaged with the source code to work on AWS: requests, sodapy. The file
+`lambda_to_socrata_util.py` that is in this repo should also be included.
+- The user must also set the following Environment Variables in their lambda
+function: SOCRATA_USERNAME, SOCRATA_PASSWORD, SOCRATA_API_KEY, SOCRATA_DATASET_ID.
+
+'''
+import boto3
+import dateutil.parser
+import json
+import logging
+import os
+import random
+import requests
+from requests.auth import HTTPBasicAuth
+requests.packages.urllib3.disable_warnings()
+from sodapy import Socrata
+import time
+
+import lambda_to_socrata_util
+
+
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+
+SOCRATA_USERNAME = os.environ['SOCRATA_USERNAME']
+SOCRATA_PASSWORD = os.environ['SOCRATA_PASSWORD']
+SOCRATA_API_KEY = os.environ['SOCRATA_API_KEY']
+SOCRATA_DATASET_ID = os.environ['SOCRATA_DATASET_ID']
+
+
+def process_bsm(raw_rec):
+    '''
+    Reads each raw BSM data record from Tampa CV Pilot and performs data transformation,
+    including:
+    1) Flatten the data structure
+    2) Rename certain fields to achieve consistency across data sets
+    3) Add additional fields to enhance usage of the data set in Socrata
+    (e.g. randomNum, coreData_position)
+
+	Parameters:
+		raw_rec: dictionary object of a single BSM record
+
+	Returns:
+		transformed dictionary object of the BSM record
+    '''
+    rename_prefix_fields = [
+        ('payload_data_coreData_', 'coreData_'),
+        ('coreData_accelSet_', 'coreData_accelset_')
+    ]
+    rename_fields = [
+        ('metadata_dataType', 'dataType'),
+        ('metadata_recordGeneratedAt', 'metadata_generatedAt'),
+        ('metadata_recordGeneratedBy', 'metadata_generatedBy'),
+        ('coreData_lat', 'coreData_position_lat'),
+        ('coreData_long', 'coreData_position_long'),
+        ('coreData_elev', 'coreData_elevation'),
```

Comment:
Shouldn't elevation also be coreData_position_elevation for consistency?

Reply:
`coreData_elev` was renamed to `coreData_elevation` to be consistent with the field name in WYDOT BSM (https://data.transportation.gov/d/9k4m-a3jc).

Reply:
Yeah, that's consistent with Wyoming but Wyoming's naming is inconsistent with the other `position` elements. Latitude, longitude and elevation are three components of a 3D position so they should all have the word `position` in their name but I understand your point. We'll leave it as is to be consistent with Wyoming.